api/git/service.go

```diff
@@ -0,0 +1,155 @@
+func NewService(ctx context.Context) *Service {
+	service := &Service{
+		shutdownCtx:  ctx,
+		azure:        NewAzureClient(REPOSITORY_CACHE_SIZE),
+		git:          NewGitClient(REPOSITORY_CACHE_SIZE),
+		timerStopped: false,
+	}
+	go service.startCacheCleanTimer(REPOSITORY_CACHE_TTL)
+	return service
+}
+
+// startCacheCleanTimer starts a timer to purge caches periodically
+func (service *Service) startCacheCleanTimer(d time.Duration) {
+	ticker := time.NewTicker(d)
+
+	for {
+		select {
+		case <-ticker.C:
+			if service.git != nil {
+				service.git.purgeCache()
+			}
+
+			if service.azure != nil {
+				service.azure.purgeCache()
+			}
+		case <-service.shutdownCtx.Done():
+			ticker.Stop()
+			service.mut.Lock()
+			service.timerStopped = true
+			service.mut.Unlock()
+			return
+		}
+	}
+}
+
+// timerHasStopped shows the CacheClean timer state with thread-safe way
+func (service *Service) timerHasStopped() bool {
+	service.mut.Lock()
+	defer service.mut.Unlock()
+	ret := service.timerStopped
+	return ret
+}
+
+// CloneRepository clones a git repository using the specified URL in the specified
+// destination folder.
+func (service *Service) CloneRepository(destination, repositoryURL, referenceName, username, password string) error {
+	options := option{
+		repositoryUrl: repositoryURL,
+		username:      username,
+		password:      password,
+		referenceName: referenceName,
+		depth:         1,
+	}
+
+	return service.cloneRepository(destination, options)
+}
+
+func (service *Service) cloneRepository(destination string, options option) error {
+	if isAzureUrl(options.repositoryUrl) {
+		return service.azure.download(context.TODO(), destination, options)
+	}
+
+	return service.git.download(context.TODO(), destination, options)
+}
+
+// LatestCommitID returns SHA1 of the latest commit of the specified reference
+func (service *Service) LatestCommitID(repositoryURL, referenceName, username, password string) (string, error) {
+	options := option{
+		repositoryUrl: repositoryURL,
+		username:      username,
+		password:      password,
+		referenceName: referenceName,
+	}
+
+	if isAzureUrl(options.repositoryUrl) {
+		return service.azure.latestCommitID(context.TODO(), options)
+	}
+
+	return service.git.latestCommitID(context.TODO(), options)
+}
+
+func (service *Service) ListRefs(repositoryURL, username, password string) ([]string, error) {
+	options := option{
+		repositoryUrl: repositoryURL,
+		username:      username,
+		password:      password,
+	}
+
+	if isAzureUrl(options.repositoryUrl) {
+		return service.azure.listRefs(context.TODO(), options)
+	}
+
+	return service.git.listRefs(context.TODO(), options)
+}
+
+func (service *Service) ListFiles(repositoryURL, referenceName, username, password string, includedExts []string) ([]string, error) {
+	options := option{
+		repositoryUrl: repositoryURL,
+		username:      username,
+		password:      password,
+		referenceName: referenceName,
+		extensions:    includedExts,
+	}
+
```

Comment:
I'd suggest moving cache lookups to the service methods. That would half the number of caches and reduce code duplication.
I'd also change `azure.listFiles` and `git.listFiles` to return full collections and do the filtering in the Service method to centralise this piece of logic.

Reply:
It is a good point. Just bring your attention that for git implementation, the type of `ref` is `[]*plumbing.Reference`, for azure, the type of `ref` is `[]string`. 
However, we still can de-duplicate the code by moving tree cache lookups and filtering in the service method